### PR TITLE
Add Foldable type class

### LINKED ISSUE #1583: Add Foldable Type Class
```
class Foldable[f] {
  def foldLeft(fn: b -> a -> b & e, s: b, xs: f[a]): b & e
  def foldRight(fn: a -> b -> b & e, s: b, xs: f[a]): b & e
}
```

```
namespace Foldable {
  def combineAll[f: Foldable, a: Monoid](fa: f[a]): a
  def count[f: Foldable](fn: a -> Bool, fa: f[a]): Int32
  def exists[f: Foldable](fn: a -> Bool, fa: f[a]): Bool
  def forall[f: Foldable](fn: a -> Bool, fa: f[a]): Bool
  def findLeft[f: Foldable](fn: a -> Bool, fa: f[a]): Option[a]
  def findRight[f: Foldable](fn: a -> Bool, fa: f[a]): Option[a]
  def foldMap[f: Foldable, m: Monoid](fn: a -> b, fa: f[a]): b
  def isEmpty[f: Foldable](fa: f[a]): Bool
  def memberOf[f: Foldable, a: Eq](x: a, fa: f[a]): Bool
  def maximum[f: Foldable, a: Order](fa: f[a]): Option[a]
  def minimumBy[f: Foldable](cmp: (a, a) -> Comparison, fa: f[a]): Option[a]
  def minimum[f: Foldable, a: Order](fa: f[a]): Option[a]
  def maximumBy[f: Foldable](cmp: (a, a) -> Comparison, fa: f[a]): Option[a]
  def reduceLeft[f: Foldable, a](fn: a -> a -> a & e, fa: f[a]): Option[a] & e
  def reduceRight[f: Foldable, a](fn: a -> a -> a & e, fa: f[a]): Option[a] & e
  def toList[f: Foldable, a](fa: f[a]): List[a]
}
```

(Feel free to double check that these signatures make sense).

Documentation for all of the above signatures can be copied and adapter from e.g. List.

Instances:
- [x] Array
- [x] Option
- [x] List
- [x] Nel

Others?

### PR DESCRIPTION
**Pull request by esbenbjerre**
Fixes #1583.

### CODE REVIEW

File: main/src/library/List.flix [hunk 1]
```diff
@@ -86,6 +86,11 @@ instance Functor[List] {
     def map(f: a -> b & e, xs: List[a]): List[b] & e = List.map(f, xs)
 }
 
+instance Foldable[List] {
+    def foldLeft(f: b -> a -> b & e, s: b, xs: List[a]): b & e = List.foldLeft(f, s, xs)
```
Comment (magnus-madsen) [inline main/src/library/List.flix:90]: I think we should write the signature uncurried as `(b, a) -> b`. Of course they are equivalent, but this is what we do in other please. I know I wrote it wrong in the ticket (copying from the api).

Comment (magnus-madsen) [reply main/src/library/List.flix:90]: This applies everywhere.

File: main/src/library/Set.flix [hunk 2]
```diff
@@ -70,6 +70,11 @@ instance Order[Set[a]] with [a : Order] {
 
 }
 
+instance Foldable[Set] {
```
Comment (magnus-madsen) [inline main/src/library/Set.flix:73]: This should not be Foldable. Later there will be an UnorderedFoldable.

File: main/src/library/Foldable.flix [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            case Some(y) => Some(Order.minBy(cmp, x, y))
+        }, None, t)
+
+    ///
+    /// Optionally finds the largest element of `t` according to the `Order` on `a`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def maximum[t :# Type -> Type : Foldable, a: Order](t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> match acc {
```
Comment (magnus-madsen) [inline main/src/library/Foldable.flix:132]: There should be some useful combinators. Maybe Option.withDefault and Option.map? or Option.flatMap or something_

Comment (esbenbjerre) [reply main/src/library/Foldable.flix:132]: I looked but it seems they all return None if one of the arguments are None. I'll check again later, there might be some combination of two functions.

**Comment by esbenbjerre** [comment]
Only `findMap` is mapping now. I'm not sure what it does and Google is of little help.

**Comment by magnus-madsen** [comment]
> Only `findMap` is missing now. I'm not sure what it does and Google is of little help.

OK that is fine. I can't find it either. Hehe.

**Comment by magnus-madsen** [comment]
I added some more instances and functions to the ticket. Don't know if you saw.

**Comment by esbenbjerre** [comment]
> I added some more instances and functions to the ticket. Don't know if you saw.

I'll add them shortly.

File: main/src/library/Foldable.flix [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    pub def memberOf[t :# Type -> Type : Foldable, a: Eq](x: a, t: t[a]): Bool =
+        Foldable.foldLeft((acc, y) -> acc or x == y, false, t)
+
+    ///
+    /// Optionally finds the smallest element of `t` according to the `Order` on `a`.
+    ///
+    /// Returns `None` if `t` is empty.
+    ///
+    pub def minimum[t :# Type -> Type : Foldable, a: Order](t: t[a]): Option[a] =
+        Foldable.foldLeft((acc, x) -> Some(Option.foldLeft((z, v) -> Order.min(z, v), x, acc)), None, t)
```
Comment (magnus-madsen) [inline main/src/library/Foldable.flix:111]: Here you can use `reduceLeft` right? :D

Same for the other minimum/maximum functions.

File: main/src/library/Foldable.flix [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    pub def reduceRight[t :# Type -> Type : Foldable, a, e](f: (a, a) -> a & e, t: t[a]): Option[a] & e =
+        foldRight((x, acc) -> Some(Option.foldLeft((v, z) -> f(v, z), x, acc)), None, t)
+
+    ///
+    /// Returns `t` as an immutable list.
+    ///
+    /// That is, the result is of the form: `t1 :: t2 :: ... :: tn :: Nil`.
+    ///
+    pub def toList[t :# Type -> Type : Foldable, a](t: t[a]): List[a] =
+        List.reverse(foldLeft((acc, x) -> x :: acc, Nil, t))
```
Comment (magnus-madsen) [inline main/src/library/Foldable.flix:163]: Or just straight-up foldRight with cons :)

File: main/src/library/Foldable.flix [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+///
+/// A type class for data structures that can be folded.
+///
+pub class Foldable[t :# Type -> Type] {
+
+    ///
+    /// Left-associative fold of a structure.
+    /// Applies `f` to a start value `s` and all elements in `t` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, t1), t2)..., tn)`.
```
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Foldable.flix:26]: It is indeed :)
Would you like me to remove it?

File: main/src/library/Foldable.flix (hunk 5 above)
Comment (esbenbjerre) [reply main/src/library/Foldable.flix:163]: B-but stack safety...

File: main/src/library/Foldable.flix (hunk 4 above)
Comment (esbenbjerre) [reply main/src/library/Foldable.flix]: Indeed! I implemented them before I saw you had added `reduce` to the list. I refactored it now.

File: main/src/library/Foldable.flix (hunk 6 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/Foldable.flix:26]: Yes, or make it make sense :D

File: main/src/library/Foldable.flix (hunk 5 above)
Comment (magnus-madsen) [reply main/src/library/Foldable.flix:163]: Hmm. But that will come later. Its a separate problem.

Comment (esbenbjerre) [reply main/src/library/Foldable.flix:163]: Roger ;)

File: main/src/library/Foldable.flix (hunk 6 above)
[the remark this answers is not in the document]
Comment (esbenbjerre) [reply main/src/library/Foldable.flix:26]: Would you mind adding support for TeX in comments? Just kidding. I'll take a look.

**Comment by esbenbjerre** [comment]
It is currently not possible to add an instance for `Map`. See https://github.com/flix/flix/issues/1583#issuecomment-762367621.

**Comment by esbenbjerre** [comment]
Otherwise I think this PR is ready.

**Comment by magnus-madsen** [comment]
Excellent!

**Pull request merged**